AK/BigIntBase.h

```diff
@@ -0,0 +1,595 @@
+        carry = __builtin_ia32_addcarryx_u64(carry, word1, word2, &output);
+        return static_cast<NativeWord>(output);
+#else
+#    if defined(AK_COMPILER_CLANG)
+        NativeWord ncarry, output;
+        if constexpr (SameAs<NativeWord, unsigned int>)
+            output = __builtin_addc(word1, word2, carry, reinterpret_cast<unsigned int*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long>)
+            output = __builtin_addcl(word1, word2, carry, reinterpret_cast<unsigned long*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long long>)
+            output = __builtin_addcll(word1, word2, carry, reinterpret_cast<unsigned long long*>(&ncarry));
+        else
+            VERIFY_NOT_REACHED();
+        carry = ncarry;
+        return output;
+#    endif
+#endif
+    }
+    // Note: This is usually too confusing for both GCC and Clang.
+    NativeWord output;
+    bool ncarry = __builtin_add_overflow(word1, word2, &output);
+    if (carry) {
+        ++output;
+        if (output == 0)
+            ncarry = true;
+    }
+    carry = ncarry;
+    return output;
+}
+
+ALWAYS_INLINE constexpr NativeWord sub_words(NativeWord word1, NativeWord word2, bool& carry)
+{
+    if (!__builtin_is_constant_evaluated()) {
+#if ARCH(X86_64)
+        unsigned long long output;
+#    if defined(AK_COMPILER_CLANG)
+        // This is named __builtin_ia32_sbb_u64 in G++. *facepalm*
+        carry = __builtin_ia32_subborrow_u64(carry, word1, word2, &output);
+#    else
+        carry = __builtin_ia32_sbb_u64(carry, word1, word2, &output);
+#    endif
+        return static_cast<NativeWord>(output);
+#else
+#    if defined(AK_COMPILER_CLANG)
+        NativeWord ncarry, output;
+        if constexpr (SameAs<NativeWord, unsigned int>)
+            output = __builtin_subc(word1, word2, carry, reinterpret_cast<unsigned int*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long>)
+            output = __builtin_subcl(word1, word2, carry, reinterpret_cast<unsigned long*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long long>)
+            output = __builtin_subcll(word1, word2, carry, reinterpret_cast<unsigned long long*>(&ncarry));
+        else
+            VERIFY_NOT_REACHED();
+        carry = ncarry;
+        return output;
+#    endif
+#endif
+    }
+    NativeWord output;
+    bool ncarry = __builtin_sub_overflow(word1, word2, &output);
+    if (carry) {
+        if (output == 0)
+            ncarry = true;
+        --output;
+    }
+    carry = ncarry;
+    return output;
+}
+
+// Calculate ((divident1 << word_size) + divident0) / divisor. Quotient should be guaranteed to fit into NativeWord.
+ALWAYS_INLINE constexpr NativeWord div_mod_words(NativeWord divident0, NativeWord divident1, NativeWord divisor, NativeWord& remainder)
+{
+#if ARCH(x86_64)
+    if (!__builtin_is_constant_evaluated()) {
+        NativeWord remainder_value, quotient;
+        asm("div %3"
```

Comment:
Even on -O3 clang and gcc both decide to turn the non-asm path in this function into a libcall to `__udivmodti4`.

Without knowing much about specific x86_64 instructions, why is `div` preferred here? I would assume that the compilers know something I don't about this particular division.

https://godbolt.org/z/vfPofbqqY

Reply:
I can say what div does
(Mobile dies not Show the context)
div always does a long division (dw / w, ie 32/16, 64/32, 128/64) and returns the modulo and dividend

Reply:
From my measurements, `div` is always faster than call to the `__udivmodti4` (https://quick-bench.com/q/xm0eK5I_NahZOf4PzxAXQxlq_WU) (however, sometimes GCC does magic and the difference becomes 1.2x and not 3x)

Reply:
My measurements from the previous (now deleted) comment were wrong.
LLVM does this exact trick with div instruction internally: https://codebrowser.dev/llvm/compiler-rt/lib/builtins/udivmodti4.c.html#udiv128by64to64, so there is no point in reimplementing it.